Initial implementation

markvincze opened the pull request:
Initial implementation of the node compactor, which removes underutilized nodes more aggressively than how GKE (and possibly other cloud implementations) would do out of the box.

This controller depends on the Cluster autoscaler running, otherwise it wouldn't work.

The algorithm is the following:
 - We check all the nodes for which `estafette.io/node-compactor-enabled` is enabled, and check if their CPU utilization is under the limit specified in `estafette.io/node-compactor-scale-down-cpu-request-ratio-limit`.
 - If we have at least as many underutilized nodes as `estafette.io/node-compactor-scale-down-required-underutilized-node-count`, then we start a scaledown.
 - When scaling down, we pick the node with the lowest utilization.
 - Removing a node consists of two steps:
   - Cordoning it
   - Deleting all of its pods
And then the Cluster autoscaler will notice that the node is not utilized, and it'll remove it automatically, so we don't actually delete nodes nor VMs ourselves.

In this first implementation we don't take pod disruption budget into account at all, but with an anti affinity rule at least we can ensure that we never delete all of the pods of a certain deployment in one go.

markvincze committed:
Add implementation

markvincze committed:
Remove temp code

JorritSalverda commented on .estafette.yaml:
```diff
@@ -0,0 +1,49 @@
+labels:
+  app: estafette-k8s-node-compactor
```
You can drop this label and use `${ESTAFETTE_GIT_NAME}` instead of `${ESTAFETTE_LABEL_APP}` throughout the manifest. This way we're sure that the container name follows the repository name. Most extensions that use ESTAFETTE_LABEL_APP if set now also fallback to ESTAFETTE_GIT_NAME if the app label isn't set.

JorritSalverda commented on .estafette.yaml:
```diff
@@ -0,0 +1,49 @@
+
+stages:
+  build:
+    image: golang:1.12.0-alpine3.9
```
Latest is `golang:1.12.6-alpine3.10`

JorritSalverda commented on .estafette.yaml:
```diff
@@ -0,0 +1,49 @@
+      GOFLAGS: -mod=vendor
+    commands:
+    - go test ./...
+    - go build -a -installsuffix cgo -ldflags "-X main.version=${ESTAFETTE_BUILD_VERSION} -X main.revision=${ESTAFETTE_GIT_REVISION} -X main.branch=${ESTAFETTE_GIT_BRANCH} -X main.buildDate=${ESTAFETTE_BUILD_DATETIME}" -o ./publish/${ESTAFETTE_LABEL_APP} .
```
Would be good to add `-X main.app=${ESTAFETTE_GIT_NAME}` and add an app variable in main.go which can then be used in logging.

JorritSalverda commented on Dockerfile:
```diff
@@ -0,0 +1,9 @@
+FROM scratch
```
If you inline the dockerfile in the manifest you don't need to look in a repository to see what it looks like, but you can see it from the estafette GUI. Only drawback is that you can't build it locally.

markvincze replied:
You mean it's possible to embed the content of the `Dockerfile` in the `.estafette.yaml`? Do we have an example of that already?

JorritSalverda replied:
https://github.com/estafette/estafette-extension-gke/blob/master/.estafette.yaml#L29-L42

JorritSalverda commented on Dockerfile:
```diff
@@ -0,0 +1,9 @@
+      description="The estafette-k8s-node-compactor is a Kubernetes controller that can remove underutilized nodes more aggressively than how the cloud provider's implementation (e.g. GKE) would do."
+
+COPY ca-certificates.crt /etc/ssl/certs/
+COPY estafette-k8s-node-compactor /
```
You can use `${ESTAFETTE_GIT_NAME}` throughout the dockerfile. I think the dev track of the docker extension does environment variable expansion now for both inline and separate dockerfiles.

JorritSalverda commented on go.mod:
```diff
@@ -0,0 +1,26 @@
+module github.com/estafette/estafette-k8s-node-compactor
```
Please run `go mod vendor` and add the `vendor` directory to the repo as well. This speeds up builds a lot and keeps us safe against repositories of dependencies being removed from the internet.

markvincze replied:
I did that already, and pushed the vendor folder to master (https://github.com/estafette/estafette-k8s-node-compactor/tree/master/vendor) so that it doesn't show up in the PR.

markvincze committed:
Remove unneeded code

markvincze committed:
Add README

markvincze committed:
Publish image from feature branch for testing

markvincze committed:
Remove unneeded env var

markvincze committed:
Call the Prom metric initialization function

markvincze committed:
Use the proper init() func name

markvincze committed:
Change branch in .estafette.yaml

stillstillinyou commented:
HELLO!! LOOKING FOR A GREAT BLOWJOB?? CALL CAROLINE +33661818188 OR VERO +33667301419!! OR CONTACT THEIR SLUTTY FRIEND @IRINAKS1 ;) THEY KNOW HOW TO SUCK WELL!! JEAN BAUDIN KNOWS THAT FOR SURE! HE ENJOYED THE CAROLINE THE MOST !! DON’T BELIVE? CALL HIM AND ASK +31657697635 XOXO

stillstillinyou commented:
HELLO!! LOOKING FOR A GREAT BLOWJOB?? CALL CAROLINE +33661818188 OR VERO +33667301419!! OR CONTACT THEIR SLUTTY FRIEND @IRINAKS1 ;) THEY KNOW HOW TO SUCK WELL!! JEAN BAUDIN KNOWS THAT FOR SURE! HE ENJOYED THE CAROLINE THE MOST !! DON’T BELIVE? CALL HIM AND ASK +31657697635 XOXO

vitorarins commented on README.md:
```diff
@@ -0,0 +1,81 @@
+
+![Nodes tightly packed with pods during peak hours.](/readme-peak-hours.png)
+
+But then during off hours are deployments scale down, so some of the pods are removed, thus the nodes might become uniformly underutilized.
```
I believe there is some typo in *... are deployments scale down*

vitorarins commented on kubernetes.yaml:
```diff
@@ -0,0 +1,69 @@
+                - "true"
+      containers:
+      - name: ${APP_NAME}
+        image: estafette/estafette-k8s-node-compactor:${GO_PIPELINE_LABEL}
```
is it possible to deploy this with estafette?

markvincze replied:
It'd probably be possible, but currently these controllers are released at Travix with GoCD, I think partly because we don't have Estafette release support for all of our clusters yet (for example PCI).

vitorarins commented on main.go:
```diff
@@ -0,0 +1,482 @@
+	return err
+}
+
+func drainPods(node *nodeInfo, k8sClient *k8s.Client) error {
```
Is this draining process the same as [kubectl drain](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#drain)?

markvincze replied:
I think `kubectl drain` is more sophisticated, but pretty much it also just deletes all the pods from the node.

JorritSalverda replied:
It would be good to check https://github.com/estafette/estafette-gke-preemptible-killer/blob/cf013fb243ab559042892406c6ec7a2671fac54d/kubernetes.go#L162, since it's a bit more advanced by filtering out daemonsets and kube-system components from getting deleted.

markvincze replied:
I implemented the filtering of DaemonSets and `kube-system` pods.

I haven't implemented the waiting for the drain to be done, because for the compactor it's not really relevant. It would be important if the draining fails and then the node is stuck in an Unschedulable state. I want to start testing the compactor and see if that ever happens.

vitorarins commented on main.go:
```diff
@@ -0,0 +1,482 @@
+		if err == nil {
+			labels.scaleDownInProgress = i
+		} else {
+			labels.scaleDownInProgress = false
```
Maybe would be nice to log the error in these cases?

armandomagalhesJEAN replied:
HELLO!! LOOKING FOR A GREAT BLOWJOB?? CALL CAROLINE +33661818188 OR VERO +33667301419!! OR CONTACT THEIR SLUTTY FRIEND @IRINAKS1 ;) THEY KNOW HOW TO SUCK WELL!! JEAN BAUDIN KNOWS THAT FOR SURE! HE ENJOYED THE CAROLINE THE MOST !! DON’T BELIVE? CALL HIM AND ASK +31657697635 or +31655935354 XOXO

vitorarins reviewed (commented):
I believe this is a great project that we need. Thank you for the work.

armandomagalhesJEAN commented:
> I believe this is a great project that we need. Thank you for the work.

yeah JEAN is also happy ;)

 HELLO!! LOOKING FOR A GREAT BLOWJOB?? CALL CAROLINE +33661818188 OR VERO +33667301419!! OR CONTACT THEIR SLUTTY FRIEND @IRINAKS1 ;) THEY KNOW HOW TO SUCK WELL!! JEAN BAUDIN KNOWS THAT FOR SURE! HE ENJOYED THE CAROLINE THE MOST !! DON’T BELIVE? CALL HIM AND ASK +31657697635 or +31655935354 XOXO

JorritSalverda commented on main.go:
```diff
@@ -0,0 +1,482 @@
+func pickUnderutilizedNodeToRemove(nodes []nodeInfo) *nodeInfo {
+	var pick *nodeInfo
+
+	for i, n := range nodes {
```
The cluster autoscaler also skips nodes with pods that have a mounted `hostPath` or `emptyDir` volume. Unless those pods have the `cluster-autoscaler.kubernetes.io/safe-to-evict: "true"` annotation.

markvincze replied:
I implemented some of the same logic:

 - We check if it has local storage
 - Check the annotation
 - Check if it's replicated

markvincze committed:
Handle properly if reading the node labels fails

markvincze committed:
README typo fixes, and reset metrics at the start of each run

markvincze committed:
Filter out DaemonSets and kube-system pods

markvincze committed:
Drop the app label and use the GIT_NAME env var

markvincze committed:
Bump the golang image version

markvincze committed:
Include an app label in all log messages

markvincze committed:
Use the GIT_NAME env var in the Dockerfile

markvincze committed:
Make node picking more sophisticated

markvincze committed:
Small fix

JorritSalverda commented on main.go:
```diff
@@ -19,14 +19,18 @@ import (
+const (
+	labelNodeCompactorEnabled                                 = "estafette.io/node-compactor-enabled"
+	labelNodeCompactorScaleDownCPURequestRatioLimit           = "estafette.io/node-compactor-scale-down-cpu-request-ratio-limit"
+	labelNodeCompactorScaleDownRequiredUnderutilizedNodeCount = "estafette.io/node-compactor-scale-down-required-underutilized-node-count"
```
What i've done in the other estafette controllers is introduce a 'state' annotation and json serialize a state object into it. It seems that at least `NodeCompactorScaleDownInProgress` is what you would call state, right?

markvincze replied:
Yep, that's the only field I'm saving as state.  
I agree it's more extensible to serialize a state object even though we have only one field at the moment. I'll change it accordingly.

markvincze committed:
Fix the way we decide if a pod prevents eviction

markvincze committed:
Only consider non-terminated pods for the stats calculations

markvincze committed:
Change build branch

markvincze committed:
Save state as a json, add env var to customize the sleep duration

markvincze committed:
Add env var for customizing the minimum node age

markvincze committed:
Improve error handling, write to annotation instead of label

markvincze committed:
Merge branch 'initial' of github.com:estafette/estafette-k8s-node-compactor into initial

markvincze committed:
Fix field casing

markvincze committed:
Retrieve the nodes after the pods

markvincze committed:
Export some more metrics

markvincze commented:
I think I've addressed all comments, and I've tested quite a lot on development and fixed a number of edge-case issues, and now it seems to be working properly.

markvincze committed:
Use a ConfigMap to store the configuration

markvincze committed:
Add info about the ConfigMap to the README

markvincze committed:
Adjust the ConfigMap, and add ConfigMap permissions to the rbac.yaml

markvincze committed:
Add get permission for configmaps

markvincze committed:
Make the minimum node age 30 minutes

markvincze commented:
Hey all,

I've been testing this and it's working fine on production.

I'd like to implement some improvements to the algorithm, but first I'll merge this PR and open a new one with the improvements so it's easier to review.
